[CORE-42] Improved Spend Reports - Display Workspace Costs for Billing Projects

kevinmarete opened the pull request:
### Jira Ticket: https://broadworkbench.atlassian.net/browse/CORE-42

### What

- This update enhances the billing project workspace list by adding spending data for each workspace. Using the [Spend Report API](https://rawls.dsde-prod.broadinstitute.org/#/billing_v2/getSpendReport), costs are displayed in three categories: `Total Spend`, `Total Storage`, and `Total Compute`. Additional features include sortable columns, date range filters, and a search bar for better data navigation. To preview this functionality enable the feature flag  `Improved Spend Reports`.

### Why

- The new display gives billing project owners visibility into workspace costs, making tracking and managing spending across projects easier. It also improves the user experience by handling missing spend reports gracefully (showing `"N/A"`) and optimizing performance for large-scale billing accounts.

### Testing strategy

- **Unit Testing:** Validate cost calculations, sorting, filtering, search logic, and error handling display as intended.
- **Manual Testing:** Test feature flag for limited access, and perform smoke tests on functionality.
- **Performance Testing:** Conduct load test with large dataset of 1900+ workspaces in billing project `general-dev-billing-account`.
- **Regression Testing:** Ensure new features do not disrupt existing workspace list functionality or UI components & tests.

  **Before (Feature-Flag=Off)**

https://github.com/user-attachments/assets/29c174c4-fdba-4ee0-9d6e-6fbcf636a294

https://github.com/user-attachments/assets/055b9ca3-c6b4-4364-9ddb-668b254bafa4

  **After (Feature-Flag=On)**

https://github.com/user-attachments/assets/58ab0cc4-ede5-4470-bc23-2f987d7fcb02

https://github.com/user-attachments/assets/3ee7b07c-b511-4e53-91ad-35d2aba1c5a9

kevinmarete committed:
Added spendReporting feature-flag

kevinmarete committed:
Added cost columns to workspace list

kevinmarete committed:
Added DateRangeFilter

kevinmarete committed:
Added SearchFilter

kevinmarete committed:
Added Filters to Workspace List

kevinmarete committed:
Updated feature-preview text

kevinmarete committed:
Integrated search on workspace list

kevinmarete committed:
Add void to metric calls

kevinmarete committed:
Added WorkspaceCategorySpendReport Type

kevinmarete committed:
Added WorkspaceCategorySpendReport to Workspace List

kevinmarete committed:
Updated errorMessage and useEffect

kevinmarete committed:
Fixed failing tests

kevinmarete committed:
Updated errorMessage to spendErrorMessage

kevinmarete committed:
Reordered filter location

kevinmarete committed:
Remove error

kevinmarete committed:
Added loading state

kevinmarete committed:
Check spendDetails

kevinmarete committed:
Wrapped Workspace tests with act

kevinmarete committed:
Added Spend Updating Status Spinner

kevinmarete committed:
Removed WorkspaceCategorySpendReport interface

kevinmarete committed:
Added optional spend attributes to BaseWorkspaceInfo

kevinmarete committed:
Removed WorkspaceCategorySpendReport interface

kevinmarete committed:
Refactored Billing Workspace to allow Ordering

kevinmarete committed:
Merge branch 'km_CORE-42_add_costs_billing_project_workspace_list' of https://github.com/DataBiosphere/terra-ui into km_CORE-42_add_costs_billing_project_workspace_list

kevinmarete committed:
update act import source

kevinmarete committed:
Added Filter tests

kevinmarete committed:
Set optional checks

kevinmarete committed:
Added Spend and Filter Workspace Tests

kevinmarete committed:
Merge from dev

admull commented:
A couple behavior comments:
1. Can we have only the table require the async load spinner? The current behavior appears to be locking control for the entire page, preventing people from navigating away while it's loading.
2. Can the default sort be by Total Spend descending instead of by Name?

kevinmarete committed:
Change defaultSort

kevinmarete committed:
Merge branch 'dev' into km_CORE-42_add_costs_billing_project_workspace_list

kevinmarete commented:
> A couple behavior comments:
> 
> 1. Can we have only the table require the async load spinner? The current behavior appears to be locking control for the entire page, preventing people from navigating away while it's loading.
> 2. Can the default sort be by Total Spend descending instead of by Name?

I have updated the sort behavior: https://github.com/DataBiosphere/terra-ui/pull/5165/commits/6e4a9a346e8c81b82c024663418afe26aa03b00c

The spinner ensures the user waits for spend data to load, preventing any assumption that it's 'N/A' by default. Letting them navigate away would defeat the purpose of fetching the data.

trholdridge commented:
> The spinner ensures the user waits for spend data to load, preventing any assumption that it's 'N/A' by default. Letting them navigate away would defeat the purpose of fetching the data.

I would agree about Adam about the ability to navigate away even without the table having loaded. The spinner can still be there, indicating that "N/A" is not the final value, but then if a user accidentally clicks into a very large project they can still click over to something else without waiting for the loading to finish.

trholdridge commented:
Were these dropdowns supposed to be eliminated as part of this ticket? I think I remember Adam mentioning that, but I might be confusing it with another one of the spend report tickets (or maybe I am misremembering the final verdict about it).

<img width="1141" alt="Screenshot 2024-11-12 at 4 27 31 PM" src="https://github.com/user-attachments/assets/e8dd48e5-4da4-4d1f-b387-62feb4c150cf">

trholdridge commented on src/billing/Workspaces/Workspaces.tsx:
```diff
@@ -9,11 +12,19 @@ import {
   isGoogleBillingProject,
 } from 'src/billing-core/models';
 import { ariaSort, HeaderRenderer } from 'src/components/table';
+import { Ajax } from 'src/libs/ajax';
```
You may want to narrow Ajax usage, similar to what Miguel has been doing with recent PRs ([example](https://github.com/DataBiosphere/terra-ui/pull/5146) if you haven't reviewed one yet). Fixing existing instances is out of scope, but if you are adding new ones it would probably be better to add the narrow versions.

kevinmarete replied:
Done: https://github.com/DataBiosphere/terra-ui/pull/5165/commits/9a4ef801a9fa8e2844b531ccff4f86a217d3fae8

trholdridge commented on src/billing/Workspaces/Workspaces.test.tsx:
```diff
@@ -193,4 +227,80 @@ describe('Workspaces', () => {
+      expect(workspaces[1]).toHaveTextContent(/\$40.00/);
+    });
+  });
 });
```
A couple more unit tests could be helpful. Some relevant tests could be that 'N/A' is displayed for costs when spend reporting is not turned on and that the date filtering field works with the spend report request. Those things work based on what I saw in the deployed site, but it would still help to spell them out in tests.

kevinmarete replied:
Done: https://github.com/DataBiosphere/terra-ui/pull/5165/commits/5e86a672d5dabf6852be908509a602031b1de70d

kevinmarete committed:
Narrow Ajax Usage

marctalbott commented on src/billing/Workspaces/Workspaces.tsx:
```diff
@@ -222,58 +269,157 @@ interface WorkspacesProps {
+      const aggregationKeys = ['Workspace~Category'];
+
+      setUpdating(true);
+      Ajax(signal)
```
We shouldn't issue this call if there aren't any workspaces in the billing project

kevinmarete replied:
Done: https://github.com/DataBiosphere/terra-ui/pull/5165/commits/2d79884a302ac2fc74ce217960aa2406b9ccfb45

kevinmarete committed:
Refactor sorting check

marctalbott commented on src/billing/Workspaces/Workspaces.tsx:
```diff
@@ -222,58 +269,157 @@ interface WorkspacesProps {
+            columnGap: '1.25rem',
+          }}
+        >
+          <DateRangeFilter
+            label='Date range'
+            rangeOptions={[7, 30, 90]}
+            defaultValue={selectedDays}
+            style={{ gridRowStart: 1, gridColumnStart: 1 }}
+            onChange={setSelectedDays}
+          />
```
Can we refactor the existing report to use this filter as well so the two reports are using the same component?

kevinmarete replied:
Done: https://github.com/DataBiosphere/terra-ui/pull/5165/commits/24aae6b17f45632e6fe9bcf108697679d735a3f5

marctalbott commented:
Can we also add a disclaimer/warning to users that total spend will include costs other than just storage and compute like we do on the existing spend report? It's confusing to see total spend be different from the sum of storage and compute costs otherwise

![Screenshot 2024-11-12 at 6 32 39 PM](https://github.com/user-attachments/assets/09474ce1-1655-4f2b-ba15-ddc5e8e7fa28)
![Screenshot 2024-11-12 at 6 32 45 PM](https://github.com/user-attachments/assets/2d9376e7-529d-49e3-82eb-ac69e6dd6165)

kevinmarete committed:
Remove spend report request if no workspaces

kevinmarete committed:
Removed expansion dropdowns

kevinmarete committed:
Removed expansion dropdowns

kevinmarete committed:
Removed spinnerOverlay

kevinmarete committed:
Removed expansion dropdowns

kevinmarete committed:
Removed unnecessary nesting

kevinmarete commented:
> > The spinner ensures the user waits for spend data to load, preventing any assumption that it's 'N/A' by default. Letting them navigate away would defeat the purpose of fetching the data.
> 
> I would agree about Adam about the ability to navigate away even without the table having loaded. The spinner can still be there, indicating that "N/A" is not the final value, but then if a user accidentally clicks into a very large project they can still click over to something else without waiting for the loading to finish.

Done: https://github.com/DataBiosphere/terra-ui/pull/5165/commits/fc056be2d7bfecd9f10a86db78f5551796071742

kevinmarete commented:
> Can we also add a disclaimer/warning to users that total spend will include costs other than just storage and compute like we do on the existing spend report? It's confusing to see total spend be different from the sum of storage and compute costs otherwise
> 
> ![Screenshot 2024-11-12 at 6 32 39 PM](https://private-user-images.githubusercontent.com/19961550/385504670-09474ce1-1655-4f2b-ba15-ddc5e8e7fa28.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzE0NTQ3NzcsIm5iZiI6MTczMTQ1NDQ3NywicGF0aCI6Ii8xOTk2MTU1MC8zODU1MDQ2NzAtMDk0NzRjZTEtMTY1NS00ZjJiLWJhMTUtZGRjNWU4ZTdmYTI4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTEyVDIzMzQzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBlNDlhODkzZmIwMDUwZjk4MTlhYjAwMzNhNTU5NjdlMWNkMmI4MmFkNjU2NjViYTEwMDM0YjYwYmYwZmViYmImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.crZvo2LNWAd8r5aTyWD12Q4Rrcy87rSwR4KkhEsSX14) ![Screenshot 2024-11-12 at 6 32 45 PM](https://private-user-images.githubusercontent.com/19961550/385504701-2d9376e7-529d-49e3-82eb-ac69e6dd6165.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzE0NTQ3NzcsIm5iZiI6MTczMTQ1NDQ3NywicGF0aCI6Ii8xOTk2MTU1MC8zODU1MDQ3MDEtMmQ5Mzc2ZTctNTI5ZC00OWUzLTgyZWItYWM2OWU2ZGQ2MTY1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTEyVDIzMzQzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWNkYzQ5MGIyNTM0NDExYjA4MmExZjBiNjQ5Nzc1MjcwNTE1MWQ1NDJkMjRiMTE0M2MxY2MxZmQzYTYyZDRjZDcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.qxtOJNK_yWzzJ8t-aur__PUrElitIntH36UjTIiXJtg)

Done: https://github.com/DataBiosphere/terra-ui/pull/5165/commits/d9c13545f4332f41c4b4292b7d3cef7088470adf

admull commented:
Noticed a small issue while testing on the PR site: if there are costs that are N/A in the table, sorting by cost descending lists N/A first. N/A should be last.

![image](https://github.com/user-attachments/assets/0162f05f-0e63-4fef-bd2c-23e758c1d08f)

kevinmarete committed:
Fixed default values to improve sorting

kevinmarete committed:
Move spinner to cover table only

kevinmarete committed:
Fix sorting of currency strings

kevinmarete committed:
Merge branch 'dev' into km_CORE-42_add_costs_billing_project_workspace_list

kevinmarete commented:
> > Were these dropdowns supposed to be eliminated as part of this ticket? I think I remember Adam mentioning that, but I might be confusing it with another one of the spend report tickets (or maybe I am misremembering the final verdict about it).
> 
> Correct. We no longer need the expansions because the google project is searchable in the updated UX
> 
> > I have removed the spinner and replaced it with a status: [fc056be](https://github.com/DataBiosphere/terra-ui/commit/fc056be2d7bfecd9f10a86db78f5551796071742)
> 
> This isn't how we typically handle loading of a component on Terra. I recommend a spinner scoped just to loading the table. If that can't be done, this experience needs to be discussed with UX.

This has now been fixed: https://github.com/DataBiosphere/terra-ui/pull/5165/commits/3f566d5e4e4e801abba3a654676d69c586d2cac5


https://github.com/user-attachments/assets/70dd0644-f0af-4c59-b686-7d34c1a4eb5a

kevinmarete commented:
> Noticed a small issue while testing on the PR site: if there are costs that are N/A in the table, sorting by cost descending lists N/A first. N/A should be last.
> 
> ![image](https://private-user-images.githubusercontent.com/13802365/385774303-0162f05f-0e63-4fef-bd2c-23e758c1d08f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzE1NTcyOTQsIm5iZiI6MTczMTU1Njk5NCwicGF0aCI6Ii8xMzgwMjM2NS8zODU3NzQzMDMtMDE2MmYwNWYtMGU2My00ZmVmLWJkMmMtMjNlNzU4YzFkMDhmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMTQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTE0VDA0MDMxNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTE4ZTczMGJhOGU3Y2IzY2RkZWM3YTRiNTc2YjAwZTU5Y2UyZTg0ZWI3YzFhNTlhODE5Mjk0ZDE2MDljMmZjMWUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.ZlOQiKW5EamJ1Jot0oBIyz5lKGswvd5sDe9Vfjjfwio)

This has been fixed: https://github.com/DataBiosphere/terra-ui/pull/5165/commits/b70cd23459e57fc14bc88622ee0edb1e928312bb


https://github.com/user-attachments/assets/63ba0837-1d2e-4574-bc81-5fba7da228ff

kevinmarete committed:
Allow undefined spend data

admull commented:
This is looking really good! Continuing to poke around, I think the "empty" state should be handled differently when there are no workspaces in the billing project. It looks a little weird with the new controls at the top. 

Can we make it look more like what the empty workspace list looks like?

![image](https://github.com/user-attachments/assets/207c07db-3ab2-4c6e-b3f4-7aa6031ec1c5)
![image](https://github.com/user-attachments/assets/dfbffb32-a961-458f-91a4-8eeba9e9114d)

kevinmarete committed:
Merged from Dev

kevinmarete committed:
Spacing empty workspace list

kevinmarete commented:
> This is looking really good! Continuing to poke around, I think the "empty" state should be handled differently when there are no workspaces in the billing project. It looks a little weird with the new controls at the top.
> 
> Can we make it look more like what the empty workspace list looks like?
> 
> ![image](https://private-user-images.githubusercontent.com/13802365/386203458-207c07db-3ab2-4c6e-b3f4-7aa6031ec1c5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzE1OTY0OTcsIm5iZiI6MTczMTU5NjE5NywicGF0aCI6Ii8xMzgwMjM2NS8zODYyMDM0NTgtMjA3YzA3ZGItM2FiMi00YzZlLWIzZjQtN2FhNjAzMWVjMWM1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMTQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTE0VDE0NTYzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTVhZWFmMWVlNDNmYmY0NWU1YTdkMDJhOWJmYzI2MGVhODJhMzJiMjA0Mzc5ODg1Nzc1ZmI2MWNkODQ0Zjg0YjgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.2A1I7ULB5HGLFcgU-XmG6A9HYzgPqf2BK0bA4dBr99k) ![image](https://private-user-images.githubusercontent.com/13802365/386204662-dfbffb32-a961-458f-91a4-8eeba9e9114d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzE1OTY0OTcsIm5iZiI6MTczMTU5NjE5NywicGF0aCI6Ii8xMzgwMjM2NS8zODYyMDQ2NjItZGZiZmZiMzItYTk2MS00NThmLTkxYTQtOGVlYmE5ZTkxMTRkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMTQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTE0VDE0NTYzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWVlMjI1MzM3ZDVhMTFkZGIxNDZkYmFlYjM3OTcwNTc4ZTNiOTlkYjYxMGQ5YjBkOWI5NmYyMmQ2YzkxOTU5NTAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.uvNVHBL4kXmo-K9cWCSexDSGCFpZI6KHKbGHRGhzm9U)

Added some space: [Spacing empty workspace list](https://github.com/DataBiosphere/terra-ui/pull/5165/commits/f41f8283b482bbfec02edb15cf28a39c1d87e8c5)

<img width="1791" alt="Screenshot 2024-11-14 at 10 28 17 AM" src="https://github.com/user-attachments/assets/8e4b0911-a82d-4f95-8662-f8c8318bf2e0">

trholdridge reviewed (approved):
Looks great! Nice work on the changes.

kevinmarete committed:
Added N/A test case
